### ISSUE TITLE
[Bug] Sometimes Monaco loads extraLibs into models which causes IntelliSense to blow up

### ISSUE THREAD

**Issue by ErkoKnoll** [microsoft/monaco-editor#2813]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Code

```typescript
fetch('https://unpkg.com/@avst-types/runtime@0.14.1/index.d.ts')
    .then(res => res.text())
    .then(dts => {
        monaco.editor.create(document.getElementById('container'), {
            value: "export const initialize: InitializeFunction = async (event, context) => {\n\n}",
            language: 'typescript',
        });

        monaco.languages.typescript.typescriptDefaults.setExtraLibs([{
            content: dts,
            filePath: 'file:///node_modules/@types/avst-types__runtime/index.d.ts'
        }])

        // monaco.editor.createModel(dts, undefined, monaco.Uri.file('file:///node_modules/@types/avst-types__runtime/index.d.ts'))
    })
```

### Actual Behavior

https://user-images.githubusercontent.com/9930079/145574787-3dcb1c66-e256-401f-a78e-2f3293b1d711.mov

As seen in the video. Initially IntelliSense works properly for `context` argument, inferring the type. But when I do SHIFT+F12 it breaks afterwards, the type then becomes `any`. It seems to be the issue of Monaco then going and loading this file of ambient types as a model (verified by debugging locally) and then TS is going to get confused because there are 2 identical files loaded both having the same types in it.

If I replace loading those types instead of using setExtraLibs but creating the model for it instead in the first place effectively preventing Monaco creating duplications, then it works correctly, but I no longer can do SHIFT+F12 to go to the reference.

### Expected Behavior

IntelliSense no breaking when Monaco loads libs added via `setExtraLibs` into modules when it needs to.

### Additional Context

When loading libs into Monaco, which is the correct API to do it these days, still using `setExtraLibs` or creating models for them? Apparently creating modes will break IntelliSense elsewhere (preventing SHIFT+F12 lookup).

**Comment by ErkoKnoll** [comment]
If I remove `file://` from model's URI path, then SHIFT+F1 starts working again:
`monaco.editor.createModel(dts, undefined, monaco.Uri.file('/node_modules/@types/avst-types__runtime/index.d.ts'))`

So my question is, what is the recommended way to registering libs then? Or what might be the difference between using `setExtraLibs` vs registering them as models? Because on the surface registering them as models seems to do the same trick, but without breaking IntelliSense.

**Comment by alexdima** [comment]
@ErkoKnoll Both extra libs and text models end up in the same TS language service. In the original sample, I believe the model and the extra lib end up with distinct URIs (`file:///node_modules/@types/avst-types__runtime/index.d.ts` vs `file:///file:///node_modules/@types/avst-types__runtime/index.d.ts`). I think having two distinct files declare the same global symbols ends up confusing the TS language service. You can use whichever method works for you, text models or extra libs. Text models have a slightly higher memory cost. The extra libs are instantiated as text models only when peeking definition or peeking references.

cc @orta if you think there's something of interest here (with the TS services becoming confused and thinking `context` is `any`).

**Comment by orta** [comment]
Yeah, that `any` could have a semantic error from typescript seeing a dupe, and then be giving `any` as a safe fallback. This is what our code in the playground for setting up [the node module its files looks like](https://github.com/microsoft/TypeScript-Website/blob/8e2598c1c43928f505c84a91a925891ed6418285/packages/sandbox/src/index.ts#L180-L189) if that helps

**Comment by jonathan-codaio** [comment]
We've been observing this exact same behavior for months, we hadn't been able to look deeply into it until now and assumed it was a mistake in our usage, but it looks like it's this same case, where something that we register only using addExtraLib in rare and seemingly non-deterministic cases gets converted into a redundant model. We are also seeing a model created for `/lib.dom.d.ts` even though we never touch anything related to this.

We still don't have a way to repro this on demand, but have some good hints hear thanks to @ErkoKnoll 's report and will update if we're able to come up with one.

**Comment by alexdima** [comment]
@jonathan-codaio This is intentional. Files used via `addExtraLib` may end up instantiated as models when using peek features. Only models can be displayed inside editors and in the example below we are peeking inside the lib. This is done [here](https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/src/typescript/languageFeatures.ts#L127). 

But when we instantiate a model from an extra lib or base lib, we make sure to use the exact same uri for both the model and the extra lib such that we can then [remove them on the web worker side](https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/src/typescript/tsWorker.ts#L66) when giving them to the language service. I think the problem in this issue occurs when feeding the same `.d.ts` content to the TS compiler under two distinct URIs. But I think we are filtering them. But it is also possible that the filtering is incomplete or something.

e.g.
![image](https://user-images.githubusercontent.com/5047891/146402642-13ef3a61-0b59-494b-af3c-b157a0530e0e.png)

**Comment by jonathan-codaio** [comment]
Thanks @alexdima ! Indeed, using Peek gives us a reliable repro finally.

I think the issue might simply be that the line that does the filtering is probably incorrect?

It's currently:

`const models = allModels.filter((uri) => !fileNameIsLib(uri)).map((uri) => uri.toString());`

but it probably should be:

`const models = allModels.map((uri) => uri.toString()).filter((uriStr) => !fileNameIsLib(uriStr));`

What's happening now is that what is passed to `fileNameIsLib()` is a `Uri` and not a string, so it always skips past this check:

https://github.com/microsoft/monaco-editor/blob/3ad48dbde4717be3249d7ac2b4451e271d77f07b/src/typescript/tsWorker.ts#L22

Though, I think there may not be an entry in `libFileMap` for these extra lib, either.

**Comment by alexdima** [comment]
@jonathan-codaio If you have a good repro case, would you please be willing to contribute a PR to improve things? I'd love for us to resolve problems in this area, maybe in the PR you can also share a simple repro case.

**Comment by jonathan-codaio** [comment]
I was trying to come up with a minimal repro in the Playground, and I think this should be sufficient, but I think I'm doing something stupid and I can't get Peek to work.

With this setup, if you Peek on a symbol from the extra lib, and then click the button, it should show that there are error markers, which is what we've been observing. Unfortunately when I do Peek, I get "No definition found", so it doesn't do the implicit model creation.

```
monaco.editor.create(document.getElementById('container'), {
	value: "import {Foo} from 'mylib';\nlet foo: Foo;",
	language: 'typescript'
});

monaco.languages.typescript.typescriptDefaults.addExtraLib(
    `
    declare module 'mylib' {
        export interface Foo {
            foo: string;
        }
    }
    `,
    'mylib.d.ts',
);  

window.logModelMarkers = function () {
    const errorMarkers = monaco.editor
      .getModelMarkers({})
      .filter(marker => marker.severity === monaco.MarkerSeverity.Error)
    console.log(errorMarkers);
};
```

```
<div id="container" style="height: 80%"></div>

<button onclick="logModelMarkers()">Log error model markers</button>
```

I can look into a PR for a fix here, can't commit that it will be soon given the holidays, but agree would like to help improve things here. Would you have a code pointer to how `libFileMap` is populated though, since I think that may need fixing here as well?

**Comment by jonathan-codaio** [comment]
(Also now that we can repro it, we were able to do a temporary workaround where when we detect the error, we just find the implicit model and `dispose()` it.)

**Comment by jonathan-codaio** [comment]
FWIW I'm decently confident this might be the fix:

```
        getScriptFileNames(): string[] {
                const allModels = this._ctx.getMirrorModels().map((model) => model.uri);
-               const models = allModels.filter((uri) => !fileNameIsLib(uri)).map((uri) => uri.toString());
+               const models = allModels
+                       .filter((uri) => !this._extraLibs[uri.path])
+                       .map((uri) => uri.toString())
+                       .filter((uriStr) => !fileNameIsLib(uriStr));
                return models.concat(Object.keys(this._extraLibs));
        }
```

**Comment by schotime** [comment]
I'm getting this issue I think when trying to load different extra-libs that are custom d.ts files per client, based on their config. It works for a little while, then decides that they should be promoted to models, then instead of using the updated client d.ts file it uses the one from the previous client. Obviously not good. 

Any ideas when this will get fixed?

**Comment by chrisdrackett** [comment]
we're getting around this by turning off this model creation entirely:

```
const originalCreateModel = monaco.editor.createModel
monaco.editor.createModel = function (value, language, uri, ...args) {
	// If Monaco tries to create a model for a file we've already added as extraLib, skip it
	if (uri?.path.includes("node_modules")) {
		return null as any
	}
	return originalCreateModel.call(this, value, language, uri, ...args)
}
```

feels a little heavy handed, but without this the typescript worker will use 4gb of memory before crashing the tab
